### PR TITLE
fix(components): [autocomplete] clicking inside closes popup (#17171)

### LINKED ISSUE #17171: [Component] [autocomplete] autocomplete closes automatically when clicking on a custom element
Bug Type: **`Component`**

## Environment

- Vue Version: `3.3.4`
- Element Plus Version: `2.3.14`
- Browser / OS: `Version 116.0.5845.187 (Official Build) (x86_64)`
- Build Tool: `Webpack`

### Related Component

- `el-autocomplete`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IlxuPHRlbXBsYXRlPlxuICA8ZWwtYXV0b2NvbXBsZXRlIFxuICAgIHBsYWNlaG9sZGVyPVwiY2xpY2sgaGVyZVwiXG4gICAgOmZldGNoLXN1Z2dlc3Rpb25zPVwicXVlcnlTZWFyY2hBc3luY1wiXG4gICAgdi1tb2RlbD1cInNlbGVjdGVkXCI+XG4gIFx0XHQ8dGVtcGxhdGUgI2RlZmF1bHQ9XCJ7IGl0ZW0gfVwiPlxuICAgICAgICA8ZGl2Pnt7aXRlbX19PGVsLWJ1dHRvbiBAY2xpY2suc3RvcD1cIigpID0+IG9uQ2xpY2tCdG4oaXRlbSlcIj5DbGljayBNZTwvZWwtYnV0dG9uPjwvZGl2PlxuXHRcdDwvdGVtcGxhdGU+XG4gIDwvZWwtYXV0b2NvbXBsZXRlPlxuXG48L3RlbXBsYXRlPlxuXG48c2NyaXB0PlxuXG5leHBvcnQgZGVmYXVsdCB7XG5cblxuICBkYXRhKCl7XG4gICAgcmV0dXJuIHtcbiAgICAgIHNlbGVjdGVkOiBudWxsLFxuICAgICAgb3B0aW9uczogW1wib3B0aW9uMVwiLCBcIm9wdGlvbjJcIl1cbiAgICB9XG4gIH0sXG5cbiAgbWV0aG9kczoge1xuICAgIHF1ZXJ5U2VhcmNoQXN5bmMocXVlcnlTdHJpbmcsIGNhbGxiYWNrKXtcbiAgICAgIGNhbGxiYWNrKHRoaXMub3B0aW9ucylcbiAgICB9LFxuICAgIG9uQ2xpY2tCdG4oaXRlbSl7XG4gICAgY29uc29sZS5sb2coXCJidXR0b24gY2xpY2tlZCBmb3IgaXRlbTogXCIsIGl0ZW0pXG4gICAgfVxuICB9XG59XG48L3NjcmlwdD5cblxuXG5cblxuXG5cblxuXG5cblxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnt9fQ==)

### Steps to reproduce

1. Click the input field
2. Click the "click me" button from the popup

### What is Expected?

Popup stays open

### What is actually happening?

Popup closes automatically

### Additional comments

This worked well in the vue2 element version.

After digging through the source code of the vue3 element I can see that the root cause of this is onClickOutside which is triggered even when clicked on a custom element inside of the popup.

Providing an option to disable onClickOutside would solve it

### PR DESCRIPTION
**Pull request by arthabus**
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

Fixes https://github.com/element-plus/element-plus/issues/17171

**TLDR**
1. In Element Vue2 it was possible to put custom elements like buttons with custom click handlers inside autocomplete items and prevent the popup from closing on click (using @click.stop on the custom element)
2. In Element Vue3 clicking such a button automatically closes the popup which is not always the desired behavior
3. This fix brings the control of the click behaviour back and aligns the behaviour with that of Element Vue2 version

**What was the fix:**

1. Inside of the `onClickOutside` there is now additional check to see if the click was inside of the popper popup. If so ignore the click
2. Additionally if the "click outside" is confirmed, it resets the `ignoreFocusEvent` and set's it to `false` to ensure popup opens as need when input is clicked the next time. 
    - Test case here: user clicks the input to open a popup -> user clicks the custom button inside of the popup that doesn't result in the popup closure (ignoreFocusEvent is true here) -> user clicks outside of the popup to close it (ignoreFocusEvent is reset) -> user clicks the input and the popup opens as needed

### CODE REVIEW

**Comment by tolking** [comment]
Thanks you for your contribution.

This issue has been fixed.
